design/weak-imports.md

```diff
@@ -0,0 +1,60 @@
+Weak Imports
+============
+
+It can be useful for WASI programs to weakly depend on certain functions in a
+WASI API.  For example, if a WASI API evolves to include a new function a
+program might want to continue to run on older systems that don't yet support
+the new function.  In this case a weak import mechanism allows the program to
+run on older systems and detect the presence of the function at runtime.
```

Comment:
Another use case worth mentioning here is APIs which aren't applicable in a given embedding, such as filesystem APIs on a device with no filesystem.